[1단계 - 페이먼트 모듈] 오거스(오명석) 미션 제출합니다.

ohgus opened the pull request:
## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - 배포 npm 링크: **[모달 배포](https://www.npmjs.com/package/@ohgus/modal-component), [페이먼츠 훅 배포](https://www.npmjs.com/package/@ohgus/payment-hooks)**
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: [StoryBook](https://68131cc0b137fee9cef1c4d3-ebxcqpiiik.chromatic.com/?path=/docs/modal--docs)

### 리뷰 요청 & 논의하고 싶은 내용

안녕하세요 카일 오거스 입니다!
리뷰이로 다시 만나게 되어 반갑습니다 😃
이번 미션은 헤일리와 페어로 진행하게 되었습니다. 처음으로 배포를 경험하면서 많은 어려움이 있었지만 재밌게 미션을 진행했습니다.
이번 미션 잘부탁드립니다 🙇

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

## 외부 변수 선언, 내부 변수 선언

useCardExpireDateValidate 훅을 구현하는 과정에서,
페어와 함께 “변수를 외부에 선언할지, 내부에 선언할지” 에 대해 깊이 논의하게 되었습니다.

논의의 출발점은 예외 처리 로직에서 필요한 현재 연도와 월 정보를 어떤 위치에 선언할 것인가였고,
이 변수는 오직 해당 훅 내부에서만 사용되며, 한 번 계산되면 다시 계산할 필요도 거의 없다는 특성이 있었습니다.

초기에는 “성능 최적화 관점에서 외부 선언이 더 적절하지 않을까?” 라는 방향도 검토했지만,

- 실제 렌더링/훅 호출에서 성능에 미치는 영향이 미미하다는 점
- 해당 값이 외부로 노출될 필요가 전혀 없다는 점
- 로컬 스코프 내 선언이 추후 유지보수 및 테스트에 유리하다는 점

등을 종합적으로 고려해, 최종적으로는 내부 선언을 선택하게 되었습니다.

다만 이 과정에서 한 가지 기준적인 의문이 생겼습니다:

> “이런 상황에서 우리는 어떤 우선순위를 기준으로 판단해야 할까?”

- 내부 선언을 기본값으로 두고, 효율이 필요한 경우 외부로 빼는 것이 더 바람직할까?
- 반대로, 외부 선언을 기본으로 하되, 의존성과 오염을 피해야 할 때 내부로 넣는 것이 맞을까?
- 아니면 상황마다 케이스별로 판단해야 하는 걸까?

이처럼 코드 구조, 의존성 범위, 성능, 최적화, 유지보수성 사이의 균형을 어디에 두어야 할지 아직 명확한 기준이 서지 않아
이 부분에 대해 카일은 어떻게 판단하시는지 궁금합니다!

**제가 정리해 본 선언 위치에 따른 장단점 (틀린 부분 있다면 지적 부탁드립니다!)**

### ✅ 내부 선언

**장점**

- 해당 훅/컴포넌트에서만 영향 → 변경 범위가 작음
- 테스트 시 의존성 주입이 쉬움
- 독립성 확보 → 재사용성이 좋아짐

**단점**

- 훅이 호출될 때마다 재계산 (성능 저하 가능)
- 여러 컴포넌트에서 같은 값을 사용할 경우 중복 계산
- 렌더링마다 메모리 재할당
- 복잡한 계산일 경우 useMemo 같은 최적화 필요

### ✅ 외부 선언

**장점**

- 모듈 로딩 시 한 번만 계산 → 퍼포먼스 효율적
- 여러 컴포넌트 간 동일한 값 공유 가능
- 무거운 연산은 재계산 없이 캐시 가능

**단점**

- 변경 시 영향 범위가 넓어짐 → 유지보수 어려움
- 의존성 오염 발생 가능 (특히 모듈 간)

**결론적으로**

단순한 최적화 vs 명확한 경계와 유지보수
이 두 가치 사이에서 어떤 것을 우선순위로 잡는 것이 더 바람직한 판단인지
카일이라면 어떤 기준으로 결정하실지 궁금합니다.


### 2) 이번 리뷰를 통해 논의하고 싶은 부분

## 모달 라이브러리 설계: 자율성과 프리셋 사이에서의 균형

이번 미션에서 모달 컴포넌트를 구현하면서, 하나의 <Modal />로 모든 역할을 처리하기보다는
Modal, ModalOverlay, ModalContent, ModalHeader, ModalBody, ModalFooter, ModalTitle, ModalCloseButton 등
작은 단위로 분리된 컴포넌트들을 제공해, 사용자가 필요한 구성 요소만 선택적으로 조합할 수 있도록 설계했습니다.

이 방식은 컴포넌트의 가독성과 유지보수성을 높이고, 유저가 상황에 맞게 더 유연하게 사용할 수 있다는 점에서 긍정적이라 판단했습니다.

하지만 구현을 진행하면서 한 가지 의문이 생겼습니다.
몇몇 컴포넌트는 단순히 자식 요소를 감싸기만 하고, 별다른 로직이나 의미가 없는 경우도 있습니다.
이처럼 사용자가 직접 구현하는 데 큰 어려움이 없는 컴포넌트까지 라이브러리에서 제공하는 것이 과연 의미 있는 추상화인지 고민하게 되었습니다.

이런 고민의 연장선에서, 다음과 같은 질문이 생깁니다:

라이브러리를 설계할 때, 사용자에게 최대한의 자율성을 부여하는 것이 중요할까요?
아니면 일정 수준의 “가이드라인” 혹은 “제한된 틀”을 제공해
라이브러리의 일관성과 사용 편의성을 우선하는 것이 바람직할까요?

특히 카일이라면,
“사용자 자율성”과 “라이브러리의 방향성과 개성” 사이에서 어떤 기준으로 균형을 잡으실지 궁금합니다.

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

ohgus committed:
docs: 기능 명세 작성

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
docs: 기능 명세 작성

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: style 라이브러리 설정

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: package.lock.json 업데이트

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
Merge remote-tracking branch 'pair-repo/step1' into step1

ohgus committed:
feat: Modal layout 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
Merge remote-tracking branch 'pair-repo/step1' into step1

kaori-killer committed:
test: 모달 layout 스토리북 작성

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
chore: package 이름 수정

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
refactor: file type 변경

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
chore: 모달 컴포넌트 타입 설정

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
Merge remote-tracking branch 'pair/step1' into step1

kaori-killer committed:
feat: Modal Header layout 구현

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

kaori-killer committed:
chore: tsconfig 파일 주석 제거

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
chore: prettierrc 설정

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
test: ModalHeader 레이아웃 스토리북 작성

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
style: reset css 추가

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: ModalTitle 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
test: ModalText 스토리북 작성

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

kaori-killer committed:
deploy: ModalText 컴포넌트 배포

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
feat: ModalOverlay 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
feat: Modal Body 레이아웃 구현

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

kaori-killer committed:
deploy: Modal Body, Layout 모듈 배포

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
refactor: Modal display flex 추가

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: ModalFooter 레이아웃 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
deploy: Modal Footer 모듈 배포

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

kaori-killer committed:
feat: Modal Close Button 구현

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
style: cursor pointer 추가

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: ModalContent 레이아웃 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: useModal hook 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: 폴더 경로 이동

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
deploy: Modal 훅 배포

Co-Authored-By: ohgus <123801385+ohgus@users.noreply.github.com>

ohgus committed:
feat: useCardNumberValidate 훅 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
refactor: util 함수 분리

- 값이 숫자인지
- 최대 길이를 넘지 않는지

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
fix: 잘못된 검증 로직 수정

최대 길이를 넘으면 false를 반환해야 하는데 반대로 true를 반환하고 있었음

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: 유효성 검즘을 통과하면 isValid, errorMessage를 초기화 한다

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: useCardCVCValidate 훅 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: useCardPasswordValidate 훅 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: useCardBrandValidate 훅 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: CardExpireDateValidate 훅 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: 폴더이름 변경

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
fix: 유효성 검증에서 걸린 경우 반환문을 통해 검증을 끝낸다

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
feat: 주어진 숫자 범위 내에 있는지 유효섬 검증하는 함수 구현

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
test: 잘못된 테스트 방식 수정

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

kaori-killer committed:
refactor: 컴포넌트 타입을 스타일 컴포넌트에서 부르는 순환 참조 문제 해결

kaori-killer committed:
docs: 컴포넌트 훅 사용 설명서 작성

kaori-killer committed:
refactor: 모달 컴포넌트 불 필요한 Props Type 제거

ohgus committed:
chore: 배포 설정

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
refactor: key, expireDate 순서 변경

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: cardNumber -> cardNumbers로 이름 변경

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: 내보내기 파일 생성

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
docs: 라이브러리 설명서 작성

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
chore: 패키지에서 로컬에서 불러오는 방식으로 변경

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
test: Modal 라이브러리 스토리 추가

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
test: ModalCloseButton 스토리 추가

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
refactor: ModalCloseButton props의 type 분리

Co-Authored-By: kaori-killer <75800958+kaori-killer@users.noreply.github.com>

ohgus committed:
docs: 사용 설명서에 pair 이름에서 내 이름으로 변경, 스토리북 링크 추가

ohgus committed:
chore: 최종 배포

- 스토리북 배포
- 라이브러리 버전, 이름 업데이트

wmakerjun commented:
@coderabbitai Analyze the commit history in this PR to identify recurring thought patterns or problem-solving flows the developer followed. Focus on practical and concise insights, not just categorizing commit types.

Instead of classifying by type (e.g., "feature" or "style"), extract **recurring structures or behavioral routines** evident in the flow of commits.

In particular, structure the analysis like this:

- **What recurring patterns** are observable? (e.g., feature → style → interaction)
- **Why were testing or refactoring efforts concentrated** at specific points?
- **What development cycle was repeated** at the custom hook level?
- **How did the development start and end**, especially regarding setup and deployment?

For each item, include a short, clear title and explain with **specific commit examples**.

Conclude with a **summary table** that distills the patterns for quick reference.

Finally, please provide the result in **Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the recurring work patterns this developer uses to solve problems.

Focus on these two key questions:

What consistent workflow or cycle does the developer follow?
– For example: feature → styling → testing. Explain with specific commit examples.

What does this pattern suggest about the developer’s mindset or strategy?
– e.g., feature-first, cleanup-later, test-driven, etc.

Structure the result with a clear title + example + interpretation, and provide the answer in Korean.

igy95 commented:
## 리뷰 요청 & 논의하고 싶은 내용

논의를 이어가기 위한 질문을 주시는 방식이 전에 비해 많이 좋아진 것 같네요 👍

#### 외부 변수 선언, 내부 변수 선언

1. 컴포넌트는 기본적으로 리렌더링으로 인해 다시 호출되는 특성을 가지기 때문에, 다시 호출됨으로써 업데이트 되어야 하는 상태, 로직 외에는 모두 컴포넌트 바깥에 선언하는 것이 맞다고 생각합니다.
2. 내부 / 외부의 기준을 컴포넌트(함수)가 아닌 컴포넌트(파일)로 본다면 함수 바깥에 의존성을 선언한다고 해도 여전히 이는 컴포넌트라는 모듈 내에 속해있기 때문에 여전히 영향 범위는 그 컴포넌트에 한정됩니다. 만일 컴포넌트 외부에 선언해두고 export까지 한다면 이것은 이 컴포넌트 외부에서도 접근 가능한 변수가 되겠죠.

```tsx
// Component.tsx

// 1.
const dependency1 = ''
// 2.
export const dependency2 = ''

export default function Component () {
  console.log(dependency1, dependency2)
}
```

igy95 commented on hooks/src/lib/index.ts:
각 인풋을 검증하는 단을 훅으로 제공하는 부분은 좋은 시도였다고 생각합니다 👍

igy95 reviewed (changes requested):
안녕하세요 오거스! 다시 만나 반갑습니다 ~ 이번 미션에서는 지금까지와 다르게 사용처와 제공처를 좀 더 분리해서 바라보기를 의도한 것으로 보이는데, 그러한 학습 목표에 맞는 고민을 공유해주셔서 좋았습니다 👍 피드백 반영 후 재요청 부탁드릴게요 !

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

ohgus committed:
chore: 훅 리턴 타입 내보내기 추가

ohgus committed:
feat: 각 훅의 리턴 타입 추가

ohgus committed:
refactor: renderHook을 beforeEach에서 한 번만 불러 사용할 수 있게 수정

ohgus committed:
chore: @ohgus/payment-hooks 최신 버전 배포

ohgus commented on hooks/src/lib/CardBrand/useCardBrandValidate.test.ts:
```diff
@@ -0,0 +1,29 @@
+describe('useCardBrandValidate', () => {
+  const CARD_BRANDS = ['신한카드', '카카오뱅크', '국민카드'];
+  it('카드 브랜드가 올바르면 isValid가 true이고 에러 메시지가 null이다.', () => {
+    const { result } = renderHook(() => useCardBrandValidate(CARD_BRANDS));
```
```ts
describe('useCardBrandValidate', () => {
  const CARD_BRANDS = ['신한카드', '카카오뱅크', '국민카드'];
  let result: { current: CardBrandValidateResult };

  beforeEach(() => {
    const rendered = renderHook(() => useCardBrandValidate(CARD_BRANDS));
    result = rendered.result;
  });

  it('카드 브랜드가 올바르면 isValid가 true이고 에러 메시지가 null이다.', () => {
    act(() => {
      result.current.validateCardBrand('신한카드');
    });

    expect(result.current.isValid).toBe(true);
    expect(result.current.errorMessage).toBeNull();
  });
```

위 방식처럼 수정할 수 있을것 같습니다!
다른 테스트도 같은 방식으로 수정했습니다!
446956f63acfabf4ace7706c877330b32a7fcb0d

ohgus committed:
style: display style 분기 처리 간소화

ohgus commented on components/src/lib/components/Modal/Modal.styled.tsx:
```diff
@@ -0,0 +1,9 @@
+export const StyledModal = styled.div<ModalProps>`
+  width: 100%;
+  height: 100%;
+  ${({ isOpen }) => (isOpen ? "display: block" : "display: none")};
```
수정했습니다!
cce168d661bf9b83b583bd1acc2598fd6e4c6060

ohgus commented on components/package.json:
```diff
@@ -18,6 +18,10 @@
+    "@emotion/css": "^11.13.5",
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
+    "@kaori-killer/modal-component": "^0.0.13",
```
넵! 
해당 패키지는 페어의 배포 패키지를 설치해 배포가 잘되었는지 테스트하기 위해서 추가된 패키지 입니다!
구현하는 동안 한 명의 패키지를 계속 배포하여 테스트했기 때문에 제 패키지가 아닌 페어의 패키지가 설치 되었습니다!

igy95 replied:
코드 diff에 비해 제가 약간 모호하게 질문을 드렸네요..!

`@kaori-killer/modal-component`가 여기에 추가되었는데, 직접 만들어주신 패키지가 맞는지 여쭤본 것이었습니다~ 동명의 패키지 이름을 프로젝트 내에서 찾아보려 했는데 보이질 않아서요!

ohgus commented on components/reset.css:
사실 저와 페어는 normalize.css에 대해서 알지 못해서 reset.css만 고려했습니다!
카일이 말씀해주셔서 둘의 장단점을 비교해봤습니다.

### reset.css

**장점**
- 완전한 초기화로 브라우저별 차이를 100%제거
- 일관된 베이스라인 확보로 섬세한 레이아웃 제어 가능

**단점**
- 모든 스타일을 지우기 때문에 다시 정의할 스타일이 많아짐
- css 파일이 비대해질 수 있음
- 커스터마이징 시 어디가 문제인지 파악이 어려울 수 있음

**사용하면 좋은 경우**
1. 전사적 디자인 시스템이나 컴포넌트 라이브러리 제작 시
2. 브라우저 기본 스타일을 전혀 활용하지 않는 커스텀 UI 작업
3. 작은 오차조차 허용하지 않는 엄격한 초기 가이드라인이 있을 때

### normalize.css

**장점**
- 브라우저 간 실제로 다른 부분만 보정해 효율적임
- 기본적인 HTML 요소의 유용한 스타일 유지
- 파일 크기가 작아 로딩 및 유지 보수에 유리

**단점**
- 브라우저 기본 스타일을 “일부” 남겨두므로 아주 세밀한 컨트롤에는 한계가 있음
- 프로젝트에 맞지 않는 기본값은 별도 오버라이드 필요

**사용하면 좋은 경우**
1. 일반 웹사이트나 빠른 프로토타이핑
2. 기본 스타일을 거의 그대로 활용하면서 브라우저 간 호환성 보장
3. 접근성과 일관된 UI 동작을 중요시할 때

### reset.css를 사용한 계기
`<h1>, <h2>, <h3>` 등의 기본 마진이 브라우저마다 달라서, 제목 간 여백을 일관되게 초기화하려 했습니다.
사실 일부의 CSS 커스텀(마진, 폰트 크기·두께)만 필요했고 많은 CSS 커스텀은 필요 없었기 때문에, 둘의 장단점을 비교하면 normalize.css로 시작해 필요한 부분만 오버라이드하는 방식도 충분히 괜찮았을 것 같습니다.

igy95 replied:
더 코멘트 드릴 부분 없을 것 같습니다 :) 현재의 이해도만으로 다음에는 합리적인 의사결정이 가능하겠네요~

ohgus commented on components/src/App.tsx:
```diff
@@ -1,10 +1,36 @@
+  ModalCloseButton,
+  ModalBody,
+  ModalFooter,
+  useModal,
```
저희가 훅을 분리한 이유는 다음과 같습니다!

### 책임 분리
- 모달 컴포넌트는 '어떻게 보일지"와 같은 UI 전용 책임만 담당하게 한다.
- `useModal` 훅은 '언제 열고 닫을지', '외부에서 상태를 어떻게 제어할지'라는 로직을 담당한다.

### 재사용성
- 훅을 통해 상태 관리 함수를 어디서든 호출할 수 있어, 버튼 클릭뿐 아니라 사용자가 원하는 이벤트 혹은 외부 API에 트리거를 연결할 수 있게 한다.

igy95 replied:
하지만 Modal 컴포넌트를 사용하는 곳에서는 useModal도 같이 사용해야 할 가능성이 높기 때문에 아래와 같은 방향도 한번 고려해보면 좋을 것 같습니다~ (필수는 아닙니다)

1. Modal 컴포넌트 선언만으로 기본적인 열림, 닫힘 기능을 가진 모달을 보여준다.
2. 외부에서 상태 조작이 필요할 경우를 고려하여 외부 상태 연결이 가능한 prop을 옵셔널하게 열어준다.

참고 - https://www.radix-ui.com/primitives/docs/components/dialog#root

ohgus commented on components/package.json:
peerDependencies 처음 알게된 부분인데 꼭 필요할 것 같습니다!

### 장점
- 중복 설치 방지: react 같은 대형 라이브러리가 두 번 번들되지 않아 빌드 번들 크기를 줄여준다.
- 단일 인스턴스 보장: 여러 버전의 react가 섞여 발생하는 훅 훼손, 컨텍스트 깨짐 같은 버그를 예방한다.

### 사용해야 하는 경우
- 내 라이브러리가 React 컴포넌트, Emotion 스타일 등 특정 런타임 api 위에서 동작할 때
- 호스트 애플리케이션에 이미 설치된 모듈을 공유해서 사용해야 할 때

따라서 다음과 같이 수정하게 될 것 같습니다!

```
"peerDependencies": {
    "react": "^19.1.0",
    "react-dom": "^19.1.0",
    "@emotion/css": "^11.13.5",
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0"
  },
```

ohgus replied:
89865cf32ecb94727b1c6e08a50d3220273f8e32
0f2350d58809f13393dffa1d1f10e12e24f0551f

peerDependencies와 devDependencies를 한 번에 적용했습니다!

ohgus commented on components/package.json:
```diff
@@ -1,7 +1,7 @@
-  "version": "0.0.1",
-  "description": "npm modules",
+  "name": "@ohgus/modal-component",
+  "version": "0.0.16",
```
카일이 저 질문을 주신 이유가 뭔지 찾아보다 `시멘틱 버저닝`이라는 것을 알게되었습니다.

혹시 `시멘틱 버저닝`에 관해서 알고 있는지 그리고 그 형식을 기준으로 어떻게 버저닝을 의도했는지 물어보신게 맞을까요?

만약 맞다면 저희는 `시멘틱 버저닝`에 대해서 몰랐습니다.
그래서 페어와 다음과 같은 배포 사이클을 기준으로 버전을 올렸습니다.

1. `기능 개발 -> 테스트 작성 -> 배포`
2. `오류 수정 -> 테스트 -> 배포`

하지만 이런 규칙들을 시멘틱 버저닝에 비교하면 저희는 어떤 상황에서도 PATCH처럼 버전을 올렸기 때문에 유저가 봤을 때 해당 버저닝이 어떤 의도의 버저닝인지 파악하기 힘들거라 예상이 됩니다.

## 시멘틱 버저닝

### 형식 규칙

```
MAJOR.MINOR.PATCH
```

- MAJOR: 호환성을 깬 변경이 있을 때 증가
- MINOR: 기존 호환성을 유지하면서 기능을 추가하거나 개선했을 때 증가
- PATCH: 호환성을 유지하면서 버그를 수정했을 때 증가

### 숫자 규칙

1. 0이상 정수만 사용한다.
2. 각 세그먼트가 증가할 때 이전 세그먼트들은 0으로 재설정한다.

- 1.4.6 -> 2.0.0 (MAJOR)
- 2.3.8 -> 2.4.0 (MINOR)
- 2.4.6 -> 2.4.7 (PATCH)

### 변경 범위

**호환성을 깨는 변경 (MAJOR)**

API 삭제, 시그니처 변경, 동작 방식 변경 등

**새로운 기능 추가 (MINOR)**

기존 API와 호환되면서 기능을 확장하는 것

**버그 수정 (PATCH)**

기능 동작을 바로잡기 위한 수정, 문서화 오류 수정 등

igy95 replied:
시맨틱 버저닝을 알고 있냐, 에 대한 질문은 아니었고 순전히 어떠한 규칙을 기준으로 버전을 올려주신 건지에 대해 궁금해서 여쭤봤습니다 ~! 말씀 주신 시맨틱 버저닝 방식은 거의 정론처럼 널리 알려지고 사용되는 규칙이지만, 때때로 이론과 맞지 않은 현실의 개발 사이클로 인해 실효성에 대한 의문을 제기한 글도 꽤 있어요.

추후에 팀 프로젝트를 관리하실 때 프로젝트 버전 관리를 어떻게 할지도 논의 사항에 포함시켜보면 좋겠네요 👍

ohgus commented on components/src/lib/components/ModalTitle/ModalTitle.tsx:
```diff
@@ -0,0 +1,20 @@
+  children,
+  fontSize,
+  fontWeight,
+  tag = "h1",
```
저희가 tag prop을 도입한 이유는 사용자에게 더 큰 자율성을 제공하기 위해서입니다.
예를 들어, 모달에 메인 타이틀`<h1>`뿐 아니라 서브 타이틀`<h2>`,` <h3>`을 추가하고 싶을 때, 태그가 고정되어 있으면 한 화면에 여러 개의 `<h1>`이 생겨 시멘틱 HTML 구조가 깨집니다.
따라서 tag를 열어두어 사용자가 원하는 헤딩 레벨을 선택할 수 있도록 설계했습니다.

추가로, 컴포넌트 이름이 ModalTitle이기 때문에 “서브 타이틀” 용도로 재사용하는 것이 맞을지 고민이 되었습니다. 하지만 헤딩 레벨마다 별도 컴포넌트를 만드는 것은 불필요한 중복이라 판단했고, 하나의 컴포넌트가 tag prop을 통해 `<h1>`부터 `<h6>`까지 유연하게 렌더링되도록 하면 코드도 간결해지고 유지보수성이 높아진다고 생각했습니다.

ohgus commented on components/src/lib/components/ModalFooter/ModalFooter.types.ts:
```diff
@@ -0,0 +1,8 @@
+
+export interface ModalFooterProps {
+  children: ReactNode;
+  direction?: "row" | "column";
+  align?: "start" | "center" | "end";
+  justify?: "start" | "center" | "end";
```
두 방향중에 선택해서 수정하라는 의미로 이해했는데 맞을까요?

## CSSProperties 타입 재사용

말씀하신 재활용이 맞을까요?

```ts
export interface ModalFooterProps
  extends Pick<
    CSSProperties,
    "flexDirection" | "alignItems" | "justifyContent"
    // padding, margin, gap 추가 고려
  > {
  children: ReactNode;
}
```

## flex 옵션 제거

### 장점
- 패키지의 책임 범위가 명확해진다. (모달 틀의 역할)
- 내부 구현이 단순해진다.

### 단점
- 반복되는 flex 레이아웃 코드를 사용자가 매번 작성해야 한다. (편의성 저하)

## 제 생각

패키지가 제공해야 할 "편의성"의 범위가 **사용자의 경험**과 **책임 분리** 사이에서 균형을 찾아야 한다고 느꼈습니다.

- 사용자의 경험: flex 설정으로 빠르게 레이아웃을 완성할 수 있다.
- 책임 분리: 라이브러리는 모달의 구조와 스타일 기본값만 관리하고, 구체적인 레이아웃은 사용자게에 위임한다.
- 
하지만 저는 여기에서 무엇을 선택하는 것이 좋을지 아직 판단하지 못했습니다.
카일이라면, 모달 라이브러리에서 어느 정도의 레이아웃 제어를 기본 제공하는 것이 적절하다고 보시나요?

igy95 replied:
> 두 방향중에 선택해서 수정하라는 의미로 이해했는데 맞을까요?

두 방향을 모두 적용했으면 좋겠다는 의미였습니다 !

> 말씀하신 재활용이 맞을까요?

맞습니당

> 카일이라면, 모달 라이브러리에서 어느 정도의 레이아웃 제어를 기본 제공하는 것이 적절하다고 보시나요?

좀 어려운 질문이네요. 왜냐하면 '적절하다'는 정도가 이 패키지를 사용할 팀의 상황에 따라 많이 바뀔 수 있기 때문입니다 ! 다만 모달 헤더, 모달 푸터와 같이 단순히 레이아웃을 조작하는데 사용할 컴포넌트를 정말 '모달'에서 제공해야하는 것이 맞는가, 를 바라보면 그건 아닌 것 같습니다.

ohgus commented on components/src/lib/components/ModalCloseButton/ModalCloseButton.tsx:
```diff
@@ -0,0 +1,13 @@
+const ModalCloseButton = ({ onClose }: ModalCloseButtonProps) => {
+  return (
+    <StyledModalCloseButton type="button" onClick={onClose}>
+      X
```
**닫기 버튼 1**
![image](https://github.com/user-attachments/assets/ffd60c38-9d6c-4cb5-9e01-37633b265f7d)
![image](https://github.com/user-attachments/assets/6a465d6c-c219-4a4d-b49e-d975db409f40)

igy95 replied:
저도 자주 부딪히는 고민 지점인데요, 우선 컴포넌트 인터페이스를 크게 나누면 아래처럼 나뉠 듯 해요.

1. 어떻게 보일지
2. 어디에 보일지
3. 어떤 기능을 가질지

여기서 3번 같은 경우에는 거의 변할 일이 없으니 (닫기 버튼을 눌렀을 때 닫는 기능말고 다른 기능이 들어가는 건 아무래도 이상하니까요) 기능에 집중한 공통 컴포넌트를 base로 만들고 1, 2번은 사용자에게 조합하는 식으로 맡겨두었다가 공통되는 패턴이 발견되는 경우 점진적으로 공통화 해볼 것 같습니다. 그렇다고 하면 아마 사용처에서는 아래와 같은 모습을 기대해볼 수 있을 듯해요

```tsx
return <Modal.Root>
  <Modal.Close>
    <Button variant={...} text={...} />
  </Modal.Close>
</Modal.Root>
```

igy95 replied:
요지는 사용자에게 '얼마큼의 귀찮음과 얼마큼의 유연성을 제공할 것인가'가 관건인데, 이 부분은 경험적으로 많이 쌓아보시면 좋을 것 같고 한가지 팁을 드리자면 처음부터 결합된 형태의 컴포넌트가 아닌 유연성을 최대한 제공하는 상태에서 조금씩 결합도를 높여가며 사용성을 높이는게 일을 여러번 반복하지 않는 길일 것 같습니다.

ohgus committed:
chore: package.json 수정

- peerDependencies 적용
- 테스트용 패키지 devDependencies로 이동
- 수정사항 배포를 위해 버저닝 수정 (PATCH)

ohgus committed:
chore: package.json 수정(hooks)

igy95 reviewed (approved):
안녕하세요 오거스~ 질문 의도를 한번 더 확인하는 부분이나 가독성 있게 코멘트를 작성해주시는 부분이 좋은 소통 경험으로 느껴졌습니다 👍 코멘트가 필요한 부분은 좀 더 추가해두었고 내부 패키지 관련해서 질문 하나 남겨두긴 했지만, 다음 단계에서 간단히 답변주시면 될 것 같아서 이만 머지하도록 할게요 :) 다음 단계에서 뵙겠습니다 !!

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`
and please skip `validateCardCVC` function i want other function review

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`
please skip the prev review

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`
you dont have to check app.tsx please check the other files

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

ohgus commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.
